schemaregistry/serde.go

```diff
@@ -0,0 +1,241 @@
+package schemaregistry
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"github.com/actgardner/gogen-avro/v10/parser"
+	"github.com/actgardner/gogen-avro/v10/resolver"
+	"github.com/actgardner/gogen-avro/v10/schema"
+	"github.com/confluentinc/confluent-kafka-go/kafka"
+)
+
+const magicByte byte = 0x0
+
+// MessageFactory is a factory function, which should return a pointer to
+// an instance into which we will unmarshal wire data.
+// For Avro, the name will be the name of the Avro type if it has one.
+// For JSON Schema, the name will be empty.
+// For Protobuf, the name will be the name of the message type.
+type MessageFactory func(subject string, name string) (interface{}, error)
+
+// Serializer represents a serializer
+type Serializer interface {
+	Configure(conf *kafka.ConfigMap, isKey bool) error
+	Serialize(topic string, msg interface{}) ([]byte, error)
+	Close()
+}
+
+// Deserializer represents a deserializer
+type Deserializer interface {
+	Configure(conf *kafka.ConfigMap, isKey bool) error
+	// Deserialize will call the MessageFactory to create an object
+	// into which we will unmarshal data.
+	Deserialize(topic string, payload []byte) (interface{}, error)
+	// DeserializeInto will unmarshal data into the given object.
+	DeserializeInto(topic string, payload []byte, msg interface{}) error
+	MessageFactory() MessageFactory
+	SetMessageFactory(factory MessageFactory)
+	Close()
+}
+
+// serde is a common instance for both the serializers and deserializers
+type serde struct {
+	client              Client
+	conf                *kafka.ConfigMap
+	isKey               bool
+	subjectNameStrategy SubjectNameStrategy
+}
+
+type serializer struct {
+	serde
+}
+
+type deserializer struct {
+	serde
+	messageFactory MessageFactory
+}
+
+// Configure configures the serde
+func (s *serde) Configure(conf *kafka.ConfigMap, isKey bool) error {
+	client, err := NewClient(conf)
+	if err != nil {
+		return err
+	}
+	s.client = client
+	s.conf = conf
+	s.isKey = isKey
+	s.subjectNameStrategy = TopicNameStrategy
+	return nil
+}
+
+// SubjectNameStrategy determines the subject for the given parameters
+type SubjectNameStrategy func(topic string, isKey bool, schema SchemaInfo) string
```

Comment:
The name of this type is the same as the name of the function that gets it. Maybe call this `SubjectNameStrategyFunc` or the method that gets it `GetSubjectNameStrategy`? (or something else more idiomatic?)

Reply:
Done